File: mobile/src/utils/noteRefs.ts
/**
 * Chat-Note 双向引用 (#2035)
 *
 * Auto-detect note titles in AI response text and enable
 * click-to-open / long-press-to-copy-wikilink behavior.
 *
 * Detection strategy: greedy longest-match with boundary checks
 * to avoid substring false positives (e.g. "React" ≠ "React Native").
 */

import { getNotes, invalidateNoteTitleCache, setOnInvalidateNoteTitleCache } from '../db';

export interface NoteRef {
  title: string;
  noteId: string;
  start: number;
  end: number;
}

// Module-level cache: title → noteId
let noteTitleCache: Map<string, string> | null = null;
/** Promise for in-flight loadNoteTitleMap() — dedup concurrent callers (#2526). */
let noteTitleMapPromise: Promise<Map<string, string>> | null = null;
/** True if invalidation happened while loading — prevents restoring stale data. */
let noteTitleMapDirty = false;
let cacheTimestamp = 0;
const CACHE_TTL_MS = 30_000;

// Register a callback so db.ts invalidations also clear this cache (#2527).
setOnInvalidateNoteTitleCache(() => {
  noteTitleCache = null;
  if (noteTitleMapPromise) {
    noteTitleMapDirty = true;
  }
  cacheTimestamp = 0;
});

/**
 * Load all non-template note titles from DB into a cached map (title → id).
 * Titles sorted by length descending for greedy longest-match.
 *
 * Dedups concurrent callers: if another call is already in-flight, subsequent
 * callers await the same promise instead of firing duplicate DB queries (#2526).
 */
export async function loadNoteTitleMap(): Promise<Map<string, string>> {
  const now = Date.now();
  if (noteTitleCache && now - cacheTimestamp < CACHE_TTL_MS) {
    return noteTitleCache;
  }
  if (noteTitleMapPromise) return noteTitleMapPromise;
  noteTitleMapPromise = _doLoadNoteTitleMap(now);
  return noteTitleMapPromise;
}

async function _doLoadNoteTitleMap(startTime: number): Promise<Map<string, string>> {
  try {
    const notes = await getNotes();
    const sorted = [...notes].sort((a, b) => b.title.length - a.title.length);
    const map = new Map<string, string>();
    for (const n of sorted) {
      if (n.title.trim()) {
        map.set(n.title, n.id);
      }
    }
    // Only cache if not invalidated while we were loading (#2526 read race)
    if (!noteTitleMapDirty) {
      noteTitleCache = map;
      cacheTimestamp = startTime;
    }
    return map;
  } finally {
    noteTitleMapPromise = null;
    noteTitleMapDirty = false;
  }
}

/** Force cache refresh on next load (e.g. after note create/delete). Also
 *  invalidates the db.ts cache to keep both caches in sync (#2527). */
export function clearNoteTitleCache(): void {
  noteTitleCache = null;
  cacheTimestamp = 0;
  if (noteTitleMapPromise) {
    noteTitleMapDirty = true;
  }
  // Invalidate the db.ts cache too, so both note-title caches stay in sync (#2527).
  invalidateNoteTitleCache();
}

/**
 * Find note references in text using greedy longest-match.
 *
 * @param text   Raw text to scan
 * @param titleMap  title → noteId map (from loadNoteTitleMap)
 * @returns Sorted array of NoteRef, empty if none
 */
export function findNoteReferences(
  text: string,
  titleMap: Map<string, string>,
): NoteRef[] {
  if (!titleMap.size || !text) return [];

  const refs: NoteRef[] = [];
  const lowerText = text.toLowerCase();  // hoisted — O(n²) → O(n) (#2396)

  for (const [title, noteId] of titleMap) {
    if (!title.trim()) continue;
    const lowerTitle = title.toLowerCase();
    let searchFrom = 0;

    while (true) {
      const idx = lowerText.indexOf(lowerTitle, searchFrom);
      if (idx === -1) break;

      // Boundary check: ensure match is not in the middle of a Latin word.
      // For ASCII-only titles, the character after the match must not be
      // a letter/digit/underscore (e.g. "React" in "Reactor" → no match).
      // For non-ASCII (CJK etc.) titles, boundary check is relaxed because
      // CJK characters don't form compound words like Latin does.
      const isAsciiOnly = /^[\x00-\x7F]+$/.test(title);
      const afterIdx = idx + title.length;
      const afterChar = afterIdx < text.length ? text[afterIdx] : ' ';
      const validAfter = !isAsciiOnly || !/[a-zA-Z0-9_]/.test(afterChar);
      if (!validAfter) {
        searchFrom = idx + 1;
        continue;
      }
      const beforeIdx = idx - 1;
      const beforeChar = beforeIdx >= 0 ? text[beforeIdx] : ' ';
      const validBefore = !isAsciiOnly || !/[a-zA-Z0-9_]/.test(beforeChar);
      if (!validBefore) {
        searchFrom = idx + 1;
        continue;
      }

      // Avoid overlapping matches (greedy longest-match already ensures
      // longer titles are checked first, but overlapping can still happen
      // with different-length titles at different positions).
      const overlapping = refs.some(
        r =>
          (idx >= r.start && idx < r.end) ||
          (r.start >= idx && r.start < idx + title.length),
      );
      if (!overlapping) {
        refs.push({ title, noteId, start: idx, end: idx + title.length });
      }
      searchFrom = idx + 1;
    }
  }

  return refs.sort((a, b) => a.start - b.start);
}

/**
 * Split a line of text into segments around note references.
 * Non-ref segments are returned with `isNoteRef: false`,
 * ref segments with `isNoteRef: true`.
 */
export function splitLineByNoteRefs(
  line: string,
  refs: NoteRef[],
): Array<{ text: string; isNoteRef: boolean; noteId?: string; title?: string }> {
  if (!refs.length) return [{ text: line, isNoteRef: false }];

  const parts: Array<{
    text: string;
    isNoteRef: boolean;
    noteId?: string;
    title?: string;
  }> = [];
  let cursor = 0;

  for (const ref of refs) {
    if (ref.start > cursor) {
      parts.push({
        text: line.slice(cursor, ref.start),
        isNoteRef: false,
      });
    }
    parts.push({
      text: ref.title,
      isNoteRef: true,
      noteId: ref.noteId,
      title: ref.title,
    });
    cursor = ref.end;
  }

  if (cursor < line.length) {
    parts.push({ text: line.slice(cursor), isNoteRef: false });
  }

  return parts;
}

File: native/VaultPilot.WinUI/Utils/NoteRefs.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace VaultPilot.WinUI.Utils;

/// <summary>
/// Note reference detection utility for Chat-Note bidirectional reference (#2035).
/// Detects both [[wikilink]] patterns and auto-detected note titles in text,
/// enabling clickable note references in AI chat responses.
/// </summary>
public static class NoteRefs
{
    // #3581: Static cache for sorted title list — avoids O(n log n) sort +
    // ToLowerInvariant on every call to FindNoteReferences.
    private static Dictionary<string, string>? _lastTitleMap;
    private static (string Lower, string Title)[]? _sortedTitlesCache;

    /// <summary>
    /// A detected note reference in text with position info.
    /// </summary>
    public readonly record struct NoteRef(string Title, string NoteId, int Start, int End);

    /// <summary>
    /// A text segment split around note references.
    /// </summary>
    public readonly record struct Segment(string Text, bool IsNoteRef, string? NoteId, string? Title);

    /// <summary>
    /// Finds [[wikilink]] patterns in text.
    /// Returns list of (Title, Start, End) for each wikilink in order of appearance.
    /// </summary>
    public static List<(string Title, int Start, int End)> FindWikilinks(string text)
    {
        var results = new List<(string Title, int Start, int End)>();
        if (string.IsNullOrEmpty(text))
            return results;

        var index = 0;
        while (index < text.Length)
        {
            var open = text.IndexOf("[[", index, StringComparison.Ordinal);
            if (open < 0)
                break;

            var close = text.IndexOf("]]", open + 2, StringComparison.Ordinal);
            if (close < 0)
                break;

            var title = text[(open + 2)..close].Trim();
            if (!string.IsNullOrEmpty(title))
            {
                results.Add((title, open, close + 2));
            }

            index = close + 2;
        }

        return results;
    }

    /// <summary>
    /// Find auto-detected note title references in text using greedy longest-match.
    /// Titles are checked longest-first to avoid substring false positives.
    /// Boundary checks prevent matching inside Latin words (e.g. "React" != "Reactor").
    ///
    /// Performance (#3581): Sorted title list is cached at the static level and
    /// rebuilt only when titleMap reference changes. Short texts (&lt;2 chars) also
    /// skip the scan entirely since the shortest cached title is 2 chars (#3932).
    /// </summary>
    /// <param name="text">Raw text to scan</param>
    /// <param name="titleMap">title -> noteId dictionary (from LoadNoteTitleMapAsync)</param>
    /// <returns>Sorted list of NoteRef, empty if none</returns>
    public static List<NoteRef> FindNoteReferences(string text, Dictionary<string, string> titleMap)
    {
        var refs = new List<NoteRef>();
        if (titleMap is null || titleMap.Count == 0 || string.IsNullOrEmpty(text))
            return refs;

        // #3581: The cache filter below keeps titles with t.Length >= 2, so a
        // 1-char text can never contain a title — skipping it is still a valid
        // optimization. 2-3 char texts must be scanned though: a 2-char title
        // (e.g. "AI") can match inside 3 chars of text (#3932).
        if (text.Length < 2)
            return refs;

        var lowerText = text.ToLowerInvariant();

        // #3581: Static cache for sorted title list — rebuilt only when the
        // titleMap reference changes (practically never within a 30s TTL).
        if (_lastTitleMap != titleMap || _sortedTitlesCache is null)
        {
            _lastTitleMap = titleMap;
            _sortedTitlesCache = titleMap.Keys
                .Where(t => !string.IsNullOrWhiteSpace(t) && t.Length >= 2)
                .OrderByDescending(t => t.Length)
                .Select(t => (Lower: t.ToLowerInvariant(), Title: t))
                .ToArray();
        }

        foreach (var (lowerTitle, title) in _sortedTitlesCache)
        {
            // #3581: Skip titles that can't possibly fit in remaining text.
            if (lowerTitle.Length > text.Length)
                continue;

            var noteId = titleMap[title];
            var searchFrom = 0;

            while (true)
            {
                var idx = lowerText.IndexOf(lowerTitle, searchFrom, StringComparison.Ordinal);
                if (idx < 0)
                    break;

                // Boundary check: for ASCII-only titles, ensure the match is not
                // in the middle of a Latin word (e.g. "React" in "Reactor" -> no match).
                // For non-ASCII (CJK etc.) titles, boundary check is relaxed because
                // CJK characters don't form compound words like Latin does.
                // #3932: idx comes from lowerText (the ToLowerInvariant copy), so
                // read boundary chars from lowerText too and use the lowercased
                // title length for the after-position. Some Unicode chars expand
                // on lowercasing (e.g. U+0130 'İ' -> "i̇"), which would misalign
                // indices against the original text.
                var isAsciiOnly = title.All(c => c <= 0x7f);
                var afterIdx = idx + lowerTitle.Length;
                var afterChar = afterIdx < lowerText.Length ? lowerText[afterIdx] : ' ';
                var validAfter = !isAsciiOnly || (!char.IsLetterOrDigit(afterChar) && afterChar != '_');
                if (!validAfter)
                {
                    searchFrom = idx + 1;
                    continue;
                }
                var beforeIdx = idx - 1;
                var beforeChar = beforeIdx >= 0 ? lowerText[beforeIdx] : ' ';
                var validBefore = !isAsciiOnly || (!char.IsLetterOrDigit(beforeChar) && beforeChar != '_');
                if (!validBefore)
                {
                    searchFrom = idx + 1;
                    continue;
                }

                // Avoid overlapping with already-found refs
                var overlaps = false;
                foreach (var existing in refs)
                {
                    if ((idx >= existing.Start && idx < existing.End) ||
                        (existing.Start >= idx && existing.Start < idx + title.Length))
                    {
                        overlaps = true;
                        break;
                    }
                }

                if (!overlaps)
                {
                    refs.Add(new NoteRef(title, noteId, idx, idx + title.Length));
                }

                searchFrom = idx + 1;
            }
        }

        refs.Sort((a, b) => a.Start.CompareTo(b.Start));
        return refs;
    }

    /// <summary>
    /// Splits a line of text into segments around note references.
    /// Non-ref segments return IsNoteRef=false; ref segments return IsNoteRef=true
    /// with the noteId and title populated.
    /// </summary>
    public static List<Segment> SplitLineByNoteRefs(string line, List<NoteRef> refs)
    {
        var segments = new List<Segment>();
        if (refs is null || refs.Count == 0)
        {
            segments.Add(new Segment(line, false, null, null));
            return segments;
        }

        var cursor = 0;
        foreach (var ref_ in refs)
        {
            if (ref_.Start > cursor)
            {
                segments.Add(new Segment(line[cursor..ref_.Start], false, null, null));
            }

            segments.Add(new Segment(ref_.Title, true, ref_.NoteId, ref_.Title));
            cursor = ref_.End;
        }

        if (cursor < line.Length)
        {
            segments.Add(new Segment(line[cursor..], false, null, null));
        }

        return segments;
    }
}
